File: scripts/import_xml.php
<!DOCTYPE HTML>
<!--
	Hielo by TEMPLATED
	templated.co @templatedco
	Released for free under the Creative Commons Attribution 3.0 license (templated.co/license)
-->
<html>
	<head>
    <?php
      include("menu.php");
      include("database_connection.php");
      $conn = new mysqli($servername, $username, $password, $dbname);
      // ["name"] refers to the filename on the client-side.
      $original_filename = $_FILES["xml_file"]["name"];
      // To get the filename (including the full path)
      // on the server-side, you need to use ["tmp_name"]:
      $xml_file = $_FILES["xml_file"]["tmp_name"];
    ?>

    <title>Import XML <?php echo $original_filename;?></title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<link rel="stylesheet" href="../assets/css/main.css" />

	</head>
	<body class="subpage">

		<!-- Header -->
			<header id="header">
        <div class="logo"><!--<a href="index.html">-->XML file imported <!--<span>by TEMPLATED</span></a> --></div>
				<a href="#menu">Menu</a>
			</header>

		<?php writeMenu(); ?>

		<!-- One -->
			<!-- <section id="One" class="wrapper style3">
				<div class="inner">
					<header class="align-center">
						<p>Eleifend vitae urna</p>
						<h2>Generic Page Template</h2>
					</header>
				</div>
			</section> -->

		<!-- Two -->
			<section id="two" class="wrapper style2">
				<div class="inner">
					<div class="box">
						<div class="content">
							<!--<header class="align-center">
								<p>maecenas sapien feugiat ex purus</p>
								<h2>Lorem ipsum dolor</h2>
							</header>-->

              <h1>XML file import</h1>

              <?php

                echo "<table><tr><th width='150px'>File</th><th>filename</th></tr>";
                echo "<tr><td>Selected file</td><td>" . $original_filename . "</td></tr>";
                echo "<tr><td>Uploaded file</td><td>" . $xml_file . "</td></tr></table>";

                $fileContent = file_get_contents($_FILES['xml_file']['tmp_name']);
                $xml = new SimpleXMLElement($fileContent);

                echo "<h2>Production order</h2>";
                // echo "<div class='container'><div class='row'><div class='twelve columns'>";
                $projectNumber = $xml->projectnumber[0];
                $productionOrder = (string) $xml->productionordernumber;
                $date = $xml->date;
                echo "Project " . $projectNumber . ", productie order " . $productionOrder . "<br>";
                echo "Date " . $date . "<br>";

                // echo "<h2>Top part</h2>";
                // echo $xml->parts->part[0]->number[0]; // get single value

                //Loop trough all parts
                echo "<h2>Part list</h2>";
                echo "<table><tr><th>pos</th><th>Number</th><th width='25%'>name</th><th>Version</th><th>Dimensions</th><th>Status</th></tr>";
                $partCount=1;
                foreach($xml->parts->part as $part)
                {
                  echo "<tr>";
                  echo "<td>" . $partCount . "</td>";
                  echo "<td>" . "<a href=../part_details.php?partNumber=" .$part->number . "&version=" .$part->version . ">" .$part->number . "</a></td>";
                  echo "<td>" . $part->name . "</td>";
                  echo "<td>" . $part->version . "</td>";
                  echo "<td>" . $part->dimension . "</td>";

                  // check if part is already in database (combination of number and version)
                  $query = "SELECT Number, Version FROM XML_demo.Parts WHERE Number='$part->number' AND Version='$part->version'";
                  $result = $conn->query($query);
                  if ($result->num_rows > 0) {
                    // Existing part found, don't add again
                    echo "<td>Existing</td>";
                  }
                  else {
                    // add part to database
                    // Get operations from XML
                    // <operations>
                    //   <operation>ZA</operation>
                    //   <operation>VP</operation>
                    // </operations>
                    $operation_1 = $part->operations->operation[0];
                    $operation_2 = $part->operations->operation[1];
                    $operation_3 = $part->operations->operation[2];
                    $operation_4 = $part->operations->operation[3];
                    $operation_5 = $part->operations->operation[4];

                    $query = "INSERT INTO XML_demo.Parts ("
                      ."Number, Name, Version, Created, Operation_1, Operation_2, Operation_3, Operation_4, Operation_5"
                      .") VALUES ("
                      ."'$part->number', '$part->name', '$part->version', '$date', '$operation_1', '$operation_2', '$operation_3', '$operation_4', '$operation_5')";
                    $result = $conn->query($query);
                    echo "<td>New</td>";
                  }
                  echo "<tr/>";
                  $partCount++;
                }
                echo "</table>";

                //Loop trough all part usage links
                echo "<h2>Usage links</h2>";
                $usageLinkCount=1;
                //echo "<div class='container'><div class='row'><div class='twelve columns'>";
                echo "<table><tr><th>pos</th><th>Parent</th><th>Version</th><th>ID</th><th>Child</th><th>Version</th><th>ID</th></tr>";
                foreach($xml->usageLinks->usageLink as $usageLink)
                {

                  $parentNumber = $usageLink->parent->number;
                  $parentVersion = $usageLink->parent->version;
                  $childNumber = $usageLink->child->number;
                  $childVersion = $usageLink->child->version;

                  // first look for parent ID
                  $query = "SELECT PartID FROM XML_demo.Parts WHERE Number='$parentNumber' AND Version='$parentVersion'";
                  $result = $conn->query($query);
                  if ($result->num_rows > 0) {
                    $parent = $result->fetch_assoc();
                    $parentID = $parent["PartID"];
                  }
                  // also look for child ID
                  $query = "SELECT PartID FROM XML_demo.Parts WHERE Number='$childNumber' AND Version='$childVersion'";
                  $result = $conn->query($query);
                  if ($result->num_rows > 0) {
                    $child = $result->fetch_assoc();
                    $childID = $child["PartID"];
                  }

                  echo "<tr><td>" . $usageLinkCount . "</td>";
                  echo "<td>" . $usageLink->parent->number . "</td>";
                  echo "<td>" . $usageLink->parent->version . "</td>";
                  echo "<td>" . $parentID . "</td>";
                  echo "<td>" . $usageLink->child->number . "</td>";
                  echo "<td>" . $usageLink->child->version . ")" . "</td>";
                  echo "<td>" . $childID . "</td></tr>";

                  // look for existing usageLink
                  $query = "SELECT PartUsageID FROM XML_demo.PartUsage WHERE parentID='$parentID' AND childID='$childID'";
                  $result = $conn->query($query);
                  if ($result->num_rows > 0) {
                    // Existing part usage found, don't add again
                    //echo "<td>Existing part</td>";
                  }
                  else {
                    //add part usage link to database
                    $query = "INSERT INTO XML_demo.PartUsage (parentID, childID, Quantity, Certificate) VALUES ('$parentID', '$childID', '$usageLink->quantity', '$usageLink->certificate')";
                    //echo "Add link to parent " . $parentID . " and child " . $childID . "<br>";
                    $result = $conn->query($query);
                    //echo "<td>new</td>";
                  }
                  $usageLinkCount++;
                }
                echo "</table>";
                //echo "</div><!--end of class 'twelve columns'</div>--></div><!--end of class 'Row'--></div>";
                // echo ($usageLinkCount-1) . " links found" . "<br>";

                echo "<h2>Add log</h2>";
                echo "<div class='container'><div class='row'><div class='twelve columns'>";
                $query = "INSERT INTO XML_demo.ImportLog (filename, CreateDate) VALUES ('$original_filename', '$date')";
                echo $query . "<br>";
                $result = $conn->query($query);
                echo "</div><!--end of class 'twelve columns'</div>--></div><!--end of class 'Row'--></div>";

              ?>




						</div>
					</div>
				</div>
			</section>

		<!-- Footer -->
			<footer id="footer">
				<div class="container">
					<ul class="icons">
						<li><a href="#" class="icon fa-twitter"><span class="label">Twitter</span></a></li>
						<li><a href="#" class="icon fa-facebook"><span class="label">Facebook</span></a></li>
						<li><a href="#" class="icon fa-instagram"><span class="label">Instagram</span></a></li>
						<li><a href="#" class="icon fa-envelope-o"><span class="label">Email</span></a></li>
					</ul>
				</div>
				<div class="copyright">
					&copy; Untitled. All rights reserved.
				</div>
			</footer>

		<!-- Scripts -->
			<script src="../assets/js/jquery.min.js"></script>
			<script src="../assets/js/jquery.scrollex.min.js"></script>
			<script src="../assets/js/skel.min.js"></script>
			<script src="../assets/js/util.js"></script>
			<script src="../assets/js/main.js"></script>

	</body>
</html>

File: scripts/Python/importXML.py
#!/usr/bin/python
import sys, getopt
import os
import mysql.connector
from xml.dom import minidom
from datetime import datetime
import time

print "----------------------------"
print "Import XML in MySQL database, start", time.strftime("%H:%M:%S", time.localtime())
print 'Number of arguments:', len(sys.argv), 'arguments.'
print 'Argument List:', str(sys.argv)

mydb = mysql.connector.connect(host="localhost", user="xml_demo", password="xml_demo", database="XML_demo")
global myCursor
myCursor=mydb.cursor()

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

def insertPart(number, name, version, dimension, operation_1, operation_2, operation_3, operation_4, operation_5):
    # print "Insert part", number, " -", name
    # Check if part exists
    query = "SELECT Number, Version FROM Parts WHERE Number='" + number + "' AND Version='" + version +"'";
    myCursor.execute(query)
    myresult = myCursor.fetchall()

    # if part doesn't exist :
    if len(myresult)==0:
        #insert in database
        createdDate = datetime.now().strftime("%d/%m/%Y %H:%M:%S") # dd/mm/YY H:M:S
        query="INSERT INTO Parts (number, name, version, dimension, operation_1, operation_2, operation_3, operation_4, operation_5, Created) " + "VALUES('" + number + "', '" + name + "', '" + version + "', '" + dimension + "', '" + operation_1 + "', '" + operation_2 + "', '" + operation_3 + "', '" + operation_4 + "', '" + operation_5 + "', '" + createdDate + "')";
        myCursor.execute(query)
        mydb.commit()

def getPartID(number, version):
    # retrieve part ID from database to use in parent-child usage link
    # print "  Searching ID of part",number,"[",version,"]"
    query = "SELECT partID FROM XML_demo.Parts WHERE Number='" + number + "' AND Version='" + version +"'";
    myCursor.execute(query)
    # print "myCursor is ",myCursor.fetchone()[0]
    partID = myCursor.fetchone()[0]
    #partID = myCursor
    # print "partID is ",partID

    if partID is not None:
        a=1
        # partID = myCursor.fetchone()[0])
        # print "    ID of part",number,"[",version,"] is", color.GREEN,partID,color.END
    else:
        print color.RED+color.BOLD,"    Part ID of ",number,color.END," not found"
        #print "    ",query
    return str(partID)

def insertPartUsage(parentNumber, parentVersion, childNumber, childVersion, quantity, certificate):
    # print "Parent "+ parentNumber+ " ["+ parentVersion+ "] of child "+ childNumber+ " ["+ childVersion+ "]"
    # Check if part usage link exists

    # print "  Check for existing usage link between",parentNumber," [",parentVersion,"] and ",childNumber," [",childVersion,"]"
    parentID = getPartID(parentNumber, parentVersion)
    childID = getPartID(childNumber, childVersion)
    query = "SELECT PartUsageID FROM PartUsage WHERE ParentID='" + parentID + "' AND ChildID='" + childID +"'";
    myCursor.execute(query)
    myresult = myCursor.fetchall()
    if len(myresult)==0:
        #insert new usage link
        query="INSERT INTO PartUsage (ParentID, ChildID, Quantity, Certificate) " + "VALUES('" + parentID + "', '" + childID + "', '" + quantity + "', '" + certificate + "')";
        # print query
        myCursor.execute(query)
        mydb.commit()

def importFile(xmlFile):
    # print "Importing ", xmlFile
    xmlDoc = minidom.parse(xmlFile)
    partList=xmlDoc.getElementsByTagName('part')
    # print "  File contains ", len(partList), " parts"
    numPart=0
    # newParts=0
    # newUsageLinks=0
    for part in partList:
        numPart=numPart+1
        partNumber = part.getElementsByTagName('number')[0].firstChild.nodeValue
        partName = part.getElementsByTagName('name')[0].firstChild.nodeValue
        partVersion = part.getElementsByTagName('version')[0].firstChild.nodeValue
        # print "  Part", numPart, " :", partNumber, " [", partVersion, "] -", partName

        # Dimensions element can by empty
        partDimension=""
        partDimensionElement = part.getElementsByTagName('dimension')[0].firstChild
        if partDimensionElement is not None:
            partDimension = part.getElementsByTagName('dimension')[0].firstChild.nodeValue

        # part Operations
        operation_1=''
        operation_2=''
        operation_3=''
        operation_4=''
        operation_5=''
        operationsList = part.getElementsByTagName('operation')
        if operationsList is not None:

            if len(operationsList)>=1:operation_1 = operationsList[0].firstChild.nodeValue
            if len(operationsList)>=2:operation_2 = operationsList[1].firstChild.nodeValue
            if len(operationsList)>=3:operation_3 = operationsList[2].firstChild.nodeValue
            if len(operationsList)>=4:operation_4 = operationsList[3].firstChild.nodeValue
            if len(operationsList)>=5:operation_5 = operationsList[4].firstChild.nodeValue
            #print "Operations :", operation_1, ",", operation_2, ",", operation_3, ",", operation_4, ",", operation_5

        # print "    Part operations : ", operation_1, ",", operation_2, ",", operation_3, ",", operation_4, ",", operation_5
        insertPart(partNumber, partName, partVersion, partDimension, operation_1, operation_2, operation_3, operation_4, operation_5)

        # print part.getElementsByTagName('number').text
    partUsageList=xmlDoc.getElementsByTagName('usageLink')
    # print len(partUsageList), " links"

    numUsage=0
    for partUsage in partUsageList:
        numUsage=numUsage+1
        quantity=""
        certificate=""
        # print "Part usage", numUsage
        parent=partUsage.getElementsByTagName('parent')[0]
        parentNumber=parent.getElementsByTagName('number')[0].firstChild.nodeValue
        parentVersion=parent.getElementsByTagName('version')[0].firstChild.nodeValue
        child=partUsage.getElementsByTagName('child')[0]
        childNumber=child.getElementsByTagName('number')[0].firstChild.nodeValue
        childVersion=child.getElementsByTagName('version')[0].firstChild.nodeValue
        quantity=partUsage.getElementsByTagName('quantity')[0].firstChild.nodeValue
        certificate=partUsage.getElementsByTagName('certificate')[0].firstChild.nodeValue
        insertPartUsage(parentNumber, parentVersion, childNumber, childVersion, quantity, certificate)

    # Add imported file to log table
    now = datetime.now().strftime("%d/%m/%Y %H:%M:%S") # dd/mm/YY H:M:S
    path, filename = os.path.split(xmlFile)
    query="INSERT INTO ImportLog (filename, createDate) " + "VALUES('" + filename + "', '" + now + "')";
    myCursor.execute(query)
    mydb.commit()
    print "  Import complete,", len(partList), " parts,",len(partUsageList), " links"


xmlDirectory = ''
try:
    opts, args = getopt.getopt(sys.argv[1:],"hd:",["ifile=","ofile="])
except getopt.GetoptError:
    print "Program to import all .xml files in a specified folder"
    print "This program requires one argument for XML files directory"
    print 'Usage : importXML.py -d <xmlDirectory>'
    sys.exit(2)
for opt, arg in opts:
    if opt == '-h':
        print 'importXML.py -d <xmlDirectory>'
        sys.exit()
    elif opt in ("-d", "--xmlDirectory"):
        xmlDirectory = arg
print 'XML directory is "', xmlDirectory

# List all XML files in this directory
numFiles=0
for file in os.listdir(xmlDirectory):
    if file.endswith(".xml"):
        numFiles=numFiles+1
        print"" #insert empty line before each new file
        print "Import file",color.BOLD + color.UNDERLINE,os.path.join(xmlDirectory, file),color.END
        importFile(os.path.join(xmlDirectory, file))
print numFiles, " files found"
print "Import end", time.strftime("%H:%M:%S", time.localtime())
